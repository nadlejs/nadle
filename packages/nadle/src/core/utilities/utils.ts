import Path from "node:path";

import c from "tinyrainbow";

import { DOT } from "./constants.js";

export function noop() {}

export function capitalize(str: string): string {
	if (str.length === 0) {
		return str;
	}

	return str.charAt(0).toUpperCase() + str.slice(1);
}

const MILLISECONDS_PER_SECOND = 1000;
const MILLISECONDS_PER_MINUTE = 60 * MILLISECONDS_PER_SECOND;
const MILLISECONDS_PER_HOUR = 60 * MILLISECONDS_PER_MINUTE;

export function formatTime(ms: number): string {
	const hours = Math.floor(ms / MILLISECONDS_PER_HOUR);
	const minutes = Math.floor((ms % MILLISECONDS_PER_HOUR) / MILLISECONDS_PER_MINUTE);
	const seconds = Math.floor((ms % MILLISECONDS_PER_MINUTE) / MILLISECONDS_PER_SECOND);
	const milliseconds = Math.floor(ms % MILLISECONDS_PER_SECOND);

	if (hours > 0) {
		if (minutes > 0) {
			return `${hours}h${minutes}m`;
		}

		return `${hours}h`;
	}

	if (minutes > 0) {
		if (seconds > 0) {
			return `${minutes}m${seconds}s`;
		}

		return `${minutes}m`;
	}

	if (seconds > 0) {
		return `${seconds}s`;
	}

	return `${milliseconds}ms`;
}

export function normalizeGlobPath(path: string) {
	if (path.startsWith(DOT)) {
		return path;
	}

	return `.${Path.sep}${path}`;
}

export function clamp(value: number, min: number, max: number): number {
	return Math.min(Math.max(value, min), max);
}

export function bindObject<T, K extends keyof T>(obj: T, keys: K[]): { [P in K]: T[P] extends (...args: any[]) => any ? T[P] : never } {
	return Object.fromEntries(
		keys.map((key) => {
			const value = obj[key];

			if (typeof value !== "function") {
				throw new Error(`Property ${String(key)} is not a function`);
			}

			return [key, value.bind(obj)];
		})
	) as any;
}

export function highlight(text: string): string {
	return c.bold(c.yellow(text));
}
