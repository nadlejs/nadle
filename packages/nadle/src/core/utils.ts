export function capitalize(str: string): string {
	if (str.length === 0) {
		return str;
	}

	return str.charAt(0).toUpperCase() + str.slice(1);
}

export function formatTime(time: number): string {
	if (time > 1000) {
		return `${(time / 1000).toFixed(2)}s`;
	}

	return `${Math.round(time)}ms`;
}

export function formatTimeString(date: Date): string {
	return date.toTimeString().split(" ")[0];
}
