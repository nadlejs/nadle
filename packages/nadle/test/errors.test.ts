import { it, expect, describe } from "vitest";

import { exec, createExec, expectFail } from "./utils.js";

describe("when register two tasks with the same name", () => {
	it("should throw error", async () => {
		await expect(() => createExec({ config: "duplicate-tasks" })`$0 hello`).rejects.toThrowError(`Task "hello" already registered`);
	});
});

describe("when passing invalid task name", () => {
	it("should throw error", async () => {
		await expectFail(() => exec`$0 unknown`);
	});
});

describe("when a task fails", () => {
	it("should report correctly", async () => {
		await expectFail(() => exec`$0 throwable`);
	});
});
