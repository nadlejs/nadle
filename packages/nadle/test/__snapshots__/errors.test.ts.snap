// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`when a task fails > should report correctly > stderr 1`] = `""`;

exports[`when a task fails > should report correctly > stdout 1`] = `
[log] <Yellow>></Yellow> Task <Bold>prepare</BoldDim> started
Preparing...
[log] <Green>✓</Green> Task <Bold>prepare</BoldDim> done in _ms
[log] <Yellow>></Yellow> Task <Bold>hello</BoldDim> started
Hello from nadle!
[log] <Green>✓</Green> Task <Bold>hello</BoldDim> done in _ms
[log] <Yellow>></Yellow> Task <Bold>throwable</BoldDim> started
[log] <Red>×</Red> Task <Bold>throwable</BoldDim> failed in _ms
[log] 
<Bold><Red>RUN FAILED</Red></BoldDim> in <Bold>_ms</BoldDim> <Dim>(<Bold>2</BoldDim><Dim> tasks executed, <Bold>1</BoldDim><Dim> task failed)</BoldDim>
`;

exports[`when passing invalid task name > should throw error > stderr 1`] = `[error] Task <Bold>unknown</BoldDim> not found`;

exports[`when passing invalid task name > should throw error > stdout 1`] = `
[log] 
<Bold><Red>RUN FAILED</Red></BoldDim> in <Bold>_ms</BoldDim> <Dim>(<Bold>0</BoldDim><Dim> task executed, <Bold>0</BoldDim><Dim> task failed)</BoldDim>
`;
